main.cpp

```diff
@@ -78,16 +61,41 @@ int main()
     cout << "Areas before sort: " << std::endl;
     printAreas(shapes);
 
-    std::sort(shapes.begin(), shapes.end(), sortByArea);
+    std::sort(shapes.begin(), shapes.end(), [](shared_ptr<Shape> first, shared_ptr<Shape> second) {
```

Comment:
generic lambda can be used here. It has auto params. Less typing and less refactoring in future :)
```cpp
[](auto first, auto second) { ... }
```